test/pylib/runner.py

```diff
@@ -166,33 +165,44 @@ def pytest_sessionstart(session: pytest.Session) -> None:
     if not session.config.getoption("--test-py-init"):
         return
 
-    init_testsuite_globals()
-    TestSuite.artifacts.add_exit_artifact(None, TestSuite.hosts.cleanup)
+    # Check if this is an xdist worker
+    is_xdist_worker = "xdist" in sys.modules and sys.modules["xdist"].is_xdist_worker(request_or_session=session)
 
-    # Run stuff just once for the pytest session even running under xdist.
-    if "xdist" not in sys.modules or not sys.modules["xdist"].is_xdist_worker(request_or_session=session):
+    # Always initialize globals in xdist workers (they run in separate processes)
+    # For the main process, only init if test.py hasn't done so already
+    if is_xdist_worker or ENVIRONMENT_PREPARED_FLAG not in os.environ:
+        init_testsuite_globals()
+        TestSuite.artifacts.add_exit_artifact(None, TestSuite.hosts.cleanup)
+
+    # Run stuff just once for the main pytest process (not in xdist workers).
+    # Only prepare the environment if it hasn't been prepared by test.py
+    if not is_xdist_worker and ENVIRONMENT_PREPARED_FLAG not in os.environ:
         temp_dir = pathlib.Path(session.config.getoption("--tmpdir")).absolute()
-        prepare_dirs(
+
+        prepare_environment(
             tempdir_base=temp_dir,
             modes=get_modes_to_run(session.config),
             gather_metrics=session.config.getoption("--gather-metrics"),
             save_log_on_success=session.config.getoption("--save-log-on-success"),
-        )
-        start_3rd_party_services(
-            tempdir_base=temp_dir,
             toxiproxy_byte_limit=session.config.getoption("--byte-limit"),
         )
 
+        # Mark that we need to clean up in pytest_unconfigure
+        session.config._needs_pytest_cleanup = True
+
 
-def pytest_sessionfinish(session: pytest.Session) -> None:
+def pytest_sessionfinish(session: pytest.Session, exitstatus: int) -> None:
     if not session.config.getoption("--test-py-init"):
         return
-    if getattr(TestSuite, "artifacts", None) is not None:
-        loop: AbstractEventLoop = asyncio.get_event_loop()
-        if not loop.is_running():
-            loop.run_until_complete(TestSuite.artifacts.cleanup_before_exit())
-        else:
-            loop.create_task(TestSuite.artifacts.cleanup_before_exit())
+
+    # Check if this is an xdist worker - workers should not clean up (only the main process should)
+    is_xdist_worker = "xdist" in sys.modules and sys.modules["xdist"].is_xdist_worker(request_or_session=session)
+    if is_xdist_worker or not getattr(session.config, '_needs_pytest_cleanup', False):
+        return
+
+    # test.py handles cleanup by itself, so we only clean up when running with pure pytest
+    if getattr(TestSuite, "artifacts", None) is not None and ENVIRONMENT_PREPARED_FLAG in os.environ:
```

Comment:
Pure pytest now does not set  ENVIRONMENT_PREPARED_FLAG. 
should be  getattr(session.config, '_needs_pytest_cleanup', False) here?